File: caustic-syntax/src/main/python/caustic/syntax/codec.py
import math

from caustic.runtime.thrift.ttypes import *


# Record kinds.
Structure = 'structure'
Attribute = 'attribute'
Reference = 'reference'


# Reserved Delimiters.
FieldDelimiter = '@'
ArrayDelimiter = ','


# Implicit Conversions.
def fmt(x):
    if x is None:
        return Empty
    elif isinstance(x, Transaction):
        return x
    elif isinstance(x, str):
        return text(x)
    elif isinstance(x, int):
        return real(x)
    elif isinstance(x, float):
        return real(x)
    elif isinstance(x, bool):
        return flag(x)
    else:
        raise TypeError('Unknown type for {}'.format(x))


# Literal Values.
def flag(x): return Transaction(literal=Literal(flag=x))
def real(x): return Transaction(literal=Literal(real=x))
def text(x): return Transaction(literal=Literal(text=x))


# Core Expressions.
def read(k): return Transaction(expression=Expression(read=Read(fmt(k))))
def write(k, v): return Transaction(expression=Expression(write=Write(fmt(k), fmt(v))))
def load(k): return Transaction(expression=Expression(load=Load(fmt(k))))
def store(n, v): return Transaction(expression=Expression(store=Store(fmt(n), fmt(v))))
def branch(c, p, f=None): return Transaction(expression=Expression(branch=Branch(fmt(c), fmt(p), fmt(f))))
def cons(a, b): return Transaction(expression=Expression(cons=Cons(fmt(a), fmt(b))))
def prefetch(k): return Transaction(expression=Expression(prefetch=Prefetch(fmt(k))))
def repeat(c, b): return Transaction(expression=Expression(repeat=Repeat(fmt(c), fmt(b))))
def rollback(m): return Transaction(expression=Expression(rollback=Rollback(fmt(m))))
def block(x, *rest): return reduce(lambda a, b: cons(a, b), rest, x)


# String Expressions.
Empty = text("")

def contains(x, y): return Transaction(expression=Expression(contains=Contains(fmt(x), fmt(y))))
def length(x): return Transaction(expression=Expression(length=Length(fmt(x))))
def concat(x, *rest): return reduce(lambda a, b: add(a, b), rest, x)
def substring(x, l, h=None): return Transaction(expression=Expression(slice=Slice(fmt(x), fmt(l), fmt(h) if h else length(fmt(x)))))
def matches(x, r): return Transaction(expression=Expression(matches=Matches(fmt(x), fmt(r))))
def indexOf(x, y): return Transaction(expression=Expression(indexOf=IndexOf(fmt(x), fmt(y))))


# Math Expressions.
Zero = real(0)
One = real(1)
Half = real(0.5)
Two = real(2)
E = real(math.e)
Pi = real(math.pi)

def add(x, y): return Transaction(expression=Expression(add=Add(fmt(x), fmt(y))))
def sub(x, y): return Transaction(expression=Expression(sub=Sub(fmt(x), fmt(y))))
def mul(x, y): return Transaction(expression=Expression(mul=Mul(fmt(x), fmt(y))))
def div(x, y): return Transaction(expression=Expression(div=Div(fmt(x), fmt(y))))
def mod(x, y): return Transaction(expression=Expression(mod=Mod(fmt(x), fmt(y))))
def pow(x, y): return Transaction(expression=Expression(pow=Pow(fmt(x), fmt(y))))
def log(x): return Transaction(expression=Expression(log=Log(fmt(x))))
def sin(x): return Transaction(expression=Expression(sin=Sin(fmt(x))))
def cos(x): return Transaction(expression=Expression(cos=Cos(fmt(x))))
def floor(x): return Transaction(expression=Expression(floor=Floor(fmt(x))))

def abs(x): return branch(less(x, Zero), sub(Zero, x), x)
def exp(x): return pow(E, x)
def tan(x): return div(sin(x), cos(x))
def cot(x): return div(cos(x), sin(x))
def sec(x): return div(One, cos(x))
def csc(x): return div(One, sin(x))
def sinh(x): return div(sub(exp(x), exp(sub(Zero, x))), Two)
def cosh(x): return div(add(exp(x), exp(sub(Zero, x))), Two)
def tanh(x): return div(sinh(x), cosh(x))
def coth(x): return div(cosh(x), sinh(x))
def sech(x): return div(One, cosh(x))
def csch(x): return div(One, sinh(x))
def sqrt(x): return pow(x, Half)
def ceil(x): return branch(equal(x, floor(x)), x, floor(x) + One)
def round(x): return branch(less(sub(x, floor(x)), Half), floor(x), ceil(x))
def trunc(x): return branch(less(x, Zero), ceil(x), floor(x))


# Comparison Expressions.
def both(x, y): return Transaction(expression=Expression(both=Both(fmt(x), fmt(y))))
def either(x, y): return Transaction(expression=Expression(either=Either(fmt(x), fmt(y))))
def negate(x): return Transaction(expression=Expression(negate=Negate(fmt(x))))
def equal(x, y): return Transaction(expression=Expression(equal=Equal(fmt(x), fmt(y))))
def less(x, y): return Transaction(expression=Expression(less=Less(fmt(x), fmt(y))))

def eq(x, y): return equal(x, y)
def ne(x, y): return negate(equal(x, y))
def lt(x, y): return less(x, y)
def le(x, y): return either(less(x, y), equal(x, y))
def gt(x, y): return negate(either(less(x, y), equal(x, y)))
def ge(x, y): return negate(less(x, y))
def xor(x, y): return both(either(x, y), negate(both(x, y)))

File: caustic-syntax/src/main/scala/caustic/syntax/Codec.scala
package caustic.syntax

import caustic.runtime.thrift

/**
 * A static library for generating Thrift transactions. Used by the generated sources to dynamically
 * build and execute transactions on the underlying database. Must be imported into all source files
 * in order for them to compile.
 */
object Codec {

  // Constants.
  val True: thrift.Transaction = flag(true)
  val False: thrift.Transaction = flag(false)
  val Zero: thrift.Transaction = real(0)
  val One: thrift.Transaction = real(1)
  val Two: thrift.Transaction = real(2)
  val Half: thrift.Transaction = real(0.5)
  val E: thrift.Transaction = real(math.E)
  val Pi: thrift.Transaction = real(math.Pi)
  val Empty: thrift.Transaction = text("")

  // Literals Values.
  def flag(x: Boolean): thrift.Transaction =
    thrift.Transaction.literal(thrift.Literal.flag(x))
  def real[T](x: T)(implicit num: Numeric[T]): thrift.Transaction =
    thrift.Transaction.literal(thrift.Literal.real(num.toDouble(x)))
  def text(x: String): thrift.Transaction =
    thrift.Transaction.literal(thrift.Literal.text(x))

  // Basic Expressions.
  def read(k: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.read(new thrift.Read(k)))
  def write(k: thrift.Transaction, v: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.write(new thrift.Write(k, v)))
  def load(n: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.load(new thrift.Load(n)))
  def store(n: thrift.Transaction, v: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.store(new thrift.Store(n, v)))
  def branch(c: thrift.Transaction, p: thrift.Transaction, f: thrift.Transaction = Empty): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.branch(new thrift.Branch(c, p, f)))
  def cons(a: thrift.Transaction, rest: thrift.Transaction*): thrift.Transaction =
    rest.foldLeft(a)((a, b) => thrift.Transaction.expression(thrift.Expression.cons(new thrift.Cons(a, b))))
  def repeat(c: thrift.Transaction, b: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.repeat(new thrift.Repeat(c, b)))
  def prefetch(k: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.prefetch(new thrift.Prefetch(k)))
  def rollback(r: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.rollback(new thrift.Rollback(r)))

  // Math Expressions.
  def add(x: thrift.Transaction, rest: thrift.Transaction*): thrift.Transaction =
    rest.foldLeft(x)((a, b) => thrift.Transaction.expression(thrift.Expression.add(new thrift.Add(a, b))))
  def sub(x: thrift.Transaction, rest: thrift.Transaction*): thrift.Transaction =
    rest.foldLeft(x)((a, b) => thrift.Transaction.expression(thrift.Expression.sub(new thrift.Sub(a, b))))
  def mul(x: thrift.Transaction, rest: thrift.Transaction*): thrift.Transaction =
    rest.foldLeft(x)((a, b) => thrift.Transaction.expression(thrift.Expression.mul(new thrift.Mul(a, b))))
  def div(x: thrift.Transaction, rest: thrift.Transaction*): thrift.Transaction =
    rest.foldLeft(x)((a, b) => thrift.Transaction.expression(thrift.Expression.div(new thrift.Div(a, b))))
  def mod(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.mod(new thrift.Mod(x, y)))
  def pow(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.pow(new thrift.Pow(x, y)))
  def log(x: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.log(new thrift.Log(x)))
  def sin(x: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.sin(new thrift.Sin(x)))
  def cos(x: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.cos(new thrift.Cos(x)))
  def floor(x: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.floor(new thrift.Floor(x)))

  def abs(x: thrift.Transaction): thrift.Transaction =
    branch(lt(x, Zero), sub(Zero, x), x)
  def exp(x: thrift.Transaction): thrift.Transaction =
    pow(E, x)
  def tan(x: thrift.Transaction): thrift.Transaction =
    div(sin(x), cos(x))
  def cot(x: thrift.Transaction): thrift.Transaction =
    div(cos(x), sin(x))
  def sec(x: thrift.Transaction): thrift.Transaction =
    div(One, cos(x))
  def csc(x: thrift.Transaction): thrift.Transaction =
    div(One, sin(x))
  def sinh(x: thrift.Transaction): thrift.Transaction =
    div(sub(exp(x), exp(sub(Zero, x))), Two)
  def cosh(x: thrift.Transaction): thrift.Transaction =
    div(add(exp(x), exp(sub(Zero, x))), Two)
  def tanh(x: thrift.Transaction): thrift.Transaction =
    div(sinh(x), cosh(x))
  def coth(x: thrift.Transaction): thrift.Transaction =
    div(cosh(x), sinh(x))
  def sech(x: thrift.Transaction): thrift.Transaction =
    div(One, cosh(x))
  def csch(x: thrift.Transaction): thrift.Transaction =
    div(One, sinh(x))
  def sqrt(x: thrift.Transaction): thrift.Transaction =
    pow(x, Half)
  def ceil(x: thrift.Transaction): thrift.Transaction =
    branch(eq(x, floor(x)), x, add(floor(x), One))
  def round(x: thrift.Transaction): thrift.Transaction =
    branch(lt(sub(x, floor(x)), Half), floor(x), ceil(x))

  // String Expressions.
  def contains(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.contains(new thrift.Contains(x, y)))
  def length(x: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.length(new thrift.Length(x)))
  def slice(x: thrift.Transaction, l: thrift.Transaction, h: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.slice(new thrift.Slice(x, l, h)))
  def slice(x: thrift.Transaction, l: thrift.Transaction): thrift.Transaction =
    slice(x, l, length(x))
  def matches(x: thrift.Transaction, r: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.matches(new thrift.Matches(x, r)))
  def indexOf(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.indexOf(new thrift.IndexOf(x, y)))

  // Logical Expressions.
  def and(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.both(new thrift.Both(x, y)))
  def or(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.either(new thrift.Either(x, y)))
  def not(x: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.negate(new thrift.Negate(x)))
  def eq(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.equal(new thrift.Equal(x, y)))
  def ne(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    not(eq(x, y))
  def le(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    or(eq(x, y), lt(x, y))
  def lt(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    thrift.Transaction.expression(thrift.Expression.less(new thrift.Less(x, y)))
  def ge(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    not(lt(x, y))
  def gt(x: thrift.Transaction, y: thrift.Transaction): thrift.Transaction =
    not(le(x, y))

}
